File: parsely/src/main/java/com/parsely/parselyandroid/FlushManager.java
package com.parsely.parselyandroid;

import java.util.Timer;
import java.util.TimerTask;

/**
 * Manager for the event flush timer.
 * <p>
 * Handles stopping and starting the flush timer. The flush timer
 * controls how often we send events to Parse.ly servers.
 */
class FlushManager {

    private final ParselyTracker parselyTracker;
    private final Timer parentTimer;
    private final long intervalMillis;
    private TimerTask runningTask;

    public FlushManager(ParselyTracker parselyTracker, Timer parentTimer, long intervalMillis) {
        this.parselyTracker = parselyTracker;
        this.parentTimer = parentTimer;
        this.intervalMillis = intervalMillis;
    }

    public void start() {
        if (runningTask != null) {
            return;
        }

        runningTask = new TimerTask() {
            public void run() {
                parselyTracker.flushEvents();
            }
        };
        parentTimer.scheduleAtFixedRate(runningTask, intervalMillis, intervalMillis);
    }

    public boolean stop() {
        if (runningTask == null) {
            return false;
        } else {
            boolean output = runningTask.cancel();
            runningTask = null;
            return output;
        }
    }

    public boolean isRunning() {
        return runningTask != null;
    }

    public long getIntervalMillis() {
        return intervalMillis;
    }
}

File: parsely/src/main/java/com/parsely/parselyandroid/FlushManager.kt
package com.parsely.parselyandroid

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Manager for the event flush timer.
 *
 *
 * Handles stopping and starting the flush timer. The flush timer
 * controls how often we send events to Parse.ly servers.
 */
internal interface FlushManager {
    fun start()
    fun stop()
    val isRunning: Boolean
    val intervalMillis: Long
}

internal class ParselyFlushManager(
    private val onFlush: () -> Unit,
    override val intervalMillis: Long,
    private val coroutineScope: CoroutineScope
) : FlushManager {
    private var job: Job? = null

    override fun start() {
        if (job?.isActive == true) return

        job = coroutineScope.launch {
            while (isActive) {
                delay(intervalMillis)
                onFlush.invoke()
            }
        }
    }

    override fun stop() {
        job?.cancel()
    }

    override val isRunning: Boolean
        get() = job?.isActive ?: false
}
